### ISSUE TITLE
model_id null prevents widget rendering in nbviewer

### ISSUE THREAD

**Issue by cornhundred** [jupyter-widgets/ipywidgets#1234]
I am modifying my widget's API (our group is building a [custom widget](https://github.com/MaayanLab/clustergrammer-widget) that used the [cookiecutter](https://github.com/jupyter-widgets/widget-cookiecutter) template), which works locally but does not work on nbviewer and I think the problem is that `model_id` is being set to `null`. I also could not find documentation on `model_id` in the docs.

Here is an nbviewer link to my notebook with two widgets made using the two different methods: [notebook link](http://nbviewer.jupyter.org/github/MaayanLab/clustergrammer-widget/blob/two_way_comm/DataFrame_Example.ipynb?flush_cache=true). Again, only the first widget renders on nbviewer, but both render locally.

Nbviewer console output tells me that embed-webpack.js is not able to validate `widgetViewObject` (line 96). I ran a debugger and looked at `widgetViewObject`, which shows

```
// the first widget renders and has a model_id
widgetViewObject
Object {model_id: "6bc8e3a09d47486e808ca929bf1c910a"}

// the second widget does not render has has null model_id
widgetViewObject
Object {model_id: null}
```

**Comment by SylvainCorlay** [comment]
Did you publish the js package on npm?

In order to be able to render your widget, nbviewer needs your javascript and fetches it from the cdn of npm.

**Comment by SylvainCorlay** [comment]
By the way, really really cool work on the clustergrammer widget.

**Comment by cornhundred** [comment]
Thanks @SylvainCorlay! 

The JS package is on [npm](https://www.npmjs.com/package/clustergrammer_widget) and I'm able to get notebooks with widgets to render on nbviewer ([notebook link](http://nbviewer.jupyter.org/github/MaayanLab/clustergrammer-widget/blob/master/Running_clustergrammer_widget.ipynb)). 

I was cleaning up the widget API and I made an alternative method to generate a widget, which ran locally but the widgets were not being rendered on nbviewer. This [notebook](http://nbviewer.jupyter.org/github/MaayanLab/clustergrammer-widget/blob/two_way_comm/DataFrame_Example.ipynb?flush_cache=true) (same link as above) shows both the old and new rendering methods, but only the first works. Viewing the console shows the validation error 

![screen shot 2017-03-28 at 1 11 31 pm](https://cloud.githubusercontent.com/assets/8352840/24417896/2d98459a-13b8-11e7-8066-e47adde362b5.png)


and checking the line, 142, in the file shows that `widgetViewObject` is not being validated:
```
	            valid = view_validate(widgetViewObject);
	            if (!valid) {
	                console.log(view_validate.errors);
	            }
```

Putting a debugger in this file (either `embed.js` or `embed-webpack.js`) and checking `widgetViewObject` shows the `null` `model_id`, which seems to be the problem. I can also see that the `model_id` is null in the local .ipynb file. I'm guessing that the `model_id` is not required to render locally but is required to render on nbviewer.

**Comment by cornhundred** [comment]
The problem was that my class, `Network`, which the widget was inheriting from 

```
class clustergrammer_widget(widgets.DOMWidget, Network):
```

was re-initializing itself when I loaded data and this was causing the error with `mode_id`. I turned off this re-initialization and I am no longer getting a `model_id` of `null` in the  notebook. I have to better understand why the initialization was causing this problem, but for now it is fixed (see notebook with old and new widget APIs working: [notebook link](http://nbviewer.jupyter.org/github/MaayanLab/clustergrammer-widget/blob/two_way_comm/Working_on_new_API.ipynb)).
